### PR TITLE
fix(TDP-4959): unmanaged exceptions

### PR DESCRIPTION
**Pull request by aleboulanger**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-4959

**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [ ] The new code does not introduce new technical issues (sonar / eslint)
- [ ] Functional tests have been performed
- [ ] Docker configuration files for config-std or config-cloud profiles are impacted

**Please check the browsers you've tested on**
- [ ] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

### CODE REVIEW

File: dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationCreate.java [hunk 1]
```diff
@@ -47,16 +52,23 @@ private PreparationCreate(Preparation preparation, String folderId) {
...
+            uri = uriBuilder.build();
+        } catch (URISyntaxException e) {
+            throw new TDPException(UNEXPECTED_EXCEPTION, e);
         }
 
         HttpPost preparationCreation = new HttpPost(uri);
 
         // Serialize preparation using configured serialization
-        preparationCreation.setHeader("Content-Type", APPLICATION_JSON_VALUE);
+        preparationCreation.setHeader(CONTENT_TYPE, APPLICATION_JSON_VALUE);
```
Comment (hmdebenque) [inline dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationCreate.java]: If you really want to clean that, use content type field in Entity creation:
`new ByteArrayEntity(preparationJSONValue, ContentType.APPLICATION_JSON.withCharset(UTF_8))`
and then remove the `setHeader` statement.

File: dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java [hunk 2]
```diff
@@ -13,45 +13,37 @@
...
-
-    /** The preparation version. */
-    private final String version;
-
     /**
      * @param id the preparation id.
      * @param version the preparation version.
      */
-    private PreparationGetContent(String id, String version) {
+    private PreparationGetContent(final String id, final String version) {
```
Comment (hmdebenque) [inline dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java]: is it really necessary to add `final` ?

File: dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java [hunk 3]
```diff
@@ -13,45 +13,37 @@
...
+import static org.talend.dataprep.BaseErrorCodes.UNEXPECTED_EXCEPTION;
+import static org.talend.dataprep.api.export.ExportParameters.SourceType.HEAD;
+import static org.talend.dataprep.command.Defaults.emptyStream;
+import static org.talend.dataprep.command.Defaults.pipeStream;
+
 /**
  * Command used to retrieve the preparation content.
  */
 @Component
 @Scope("request")
```
Comment (hmdebenque) [inline dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java]: `@Scope(BeanDefinition.SCOPE_PROTOTYPE)` 4ever

File: dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java [hunk 4]
```diff
@@ -77,10 +68,11 @@ private PreparationGetContent(String id, String version, ExportParameters.Source
                 post.setEntity(new StringEntity(parametersAsString, ContentType.APPLICATION_JSON));
                 return post;
             } catch (Exception e) {
-                throw new TDPException(APIErrorCodes.UNABLE_TO_RETRIEVE_PREPARATION_CONTENT, e);
+                throw new TDPException(UNEXPECTED_EXCEPTION, e);
```
Comment (hmdebenque) [inline dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java]: Great, it is not a normal error to break here :+1:

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java [hunk 5]
```diff
@@ -1348,6 +1310,60 @@ public void test_add_preparation_TDP3927() throws Exception {
...
+
+        // when trying to get the content of the preparation with an invalid version value
+        String invalidVersionId = "%00";
+        Response response = given().queryParam("version", invalidVersionId) //
+                .expect().statusCode(500) //
+                .log().ifError() //
+                .get("/api/preparations/{preparationId}/content", preparationId);
+
+        // assertions
+        assertTrue(response.jsonPath().getString("context.cause.code").endsWith(UNABLE_TO_READ_PREPARATION.getCode()));
```
Comment (hmdebenque) [inline dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: Could use mapping to TdpExceptionDto

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java [hunk 6]
```diff
@@ -1348,6 +1310,60 @@ public void test_add_preparation_TDP3927() throws Exception {
...
+        Response response = given().contentType(ContentType.JSON)
+                .body("{ \"name\": \"foo\", \"dataSetId\": \"" + dataSetId + "\"}") //
+                .queryParam("folder", invalidFolderId) //
+                .expect() //
+                .statusCode(500) //
+                .log().ifError() //
+                .when() //
+                .post("/api/preparations");
+        // assertions
+        JsonPath jsonPath = response.jsonPath();
```
Comment (hmdebenque) [inline dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: same

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java [hunk 7]
```diff
@@ -1348,6 +1310,60 @@ public void test_add_preparation_TDP3927() throws Exception {
...
+        Response response = given().contentType(ContentType.JSON)
+                .body("{ \"name\": \"foo\", \"dataSetId\": \"" + invalidDataSetId + "\"}") //
+                .queryParam("folder", "5a549eea1235ef6ee90e2096") //
+                .expect() //
+                .statusCode(400) //
+                .log().ifError() //
+                .when() //
+                .post("/api/preparations");
+        // assertions
+        assertTrue(response.jsonPath().getString("cause.code").endsWith(DATASET_DOES_NOT_EXIST.getCode()));
```
Comment (hmdebenque) [inline dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: same here

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java [hunk 8]
```diff
@@ -91,6 +60,41 @@
 import com.jayway.restassured.path.json.JsonPath;
 import com.jayway.restassured.response.Response;
 
+import static com.jayway.restassured.RestAssured.expect;
```
Comment (hmdebenque) [inline dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: Did you really need to reformat the whole file ?

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java [hunk 9]
```diff
@@ -1348,6 +1310,60 @@ public void test_add_preparation_TDP3927() throws Exception {
...
+        // given an existing dataSet
+        String dataSetId = testClient.createDataset("dataset/dataset.csv", "dataset");
+
+        // when using an invalid folder id to create a preparation on the existing dataSet
+        String invalidFolderId = "invalid-folder-id";
+        Response response = given().contentType(ContentType.JSON)
+                .body("{ \"name\": \"foo\", \"dataSetId\": \"" + dataSetId + "\"}") //
+                .queryParam("folder", invalidFolderId) //
+                .expect() //
+                .statusCode(500) //
```
Comment (hmdebenque) [inline dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: folder does not exist should result in a 4XX status code error

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java [hunk 10]
```diff
@@ -1348,6 +1310,60 @@ public void test_add_preparation_TDP3927() throws Exception {
...
 
+    @Test
+    public void shouldNotAcceptInvalidVersionId_TDP_4959() throws IOException {
+        // given a preparation
+        final String preparationId = testClient.createPreparationFromFile("dataset/dataset.csv", "foo", home.getId());
+
+        // when trying to get the content of the preparation with an invalid version value
+        String invalidVersionId = "%00";
+        Response response = given().queryParam("version", invalidVersionId) //
+                .expect().statusCode(500) //
```
Comment (hmdebenque) [inline dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: version ID does not exists : 4XX error

File: dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationCreate.java (hunk 1 above)
Comment (aleboulanger) [reply dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationCreate.java]: I don't want to change the behavior here but just use existing constant value.

File: dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java (hunk 2 above)
Comment (aleboulanger) [reply dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java]: No

File: dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java (hunk 3 above)
Comment (aleboulanger) [reply dataprep-api/src/main/java/org/talend/dataprep/api/service/command/preparation/PreparationGetContent.java]: 👌

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java (hunk 5 above)
Comment (aleboulanger) [reply dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: ok

File: dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java (hunk 8 above)
Comment (aleboulanger) [reply dataprep-api/src/test/java/org/talend/dataprep/api/service/PreparationAPITest.java]: side effect of IDE auto-reformat, according to Talend java formatter - so i guess we can keep it

**Pull request merged**